**Assess multiple sequence alignments**
---------------------------------

**Example Plots:**

The R script ***Assessment_Figures_Basic_v2.R*** was used to create the histograms and scatterplots
contained here. These files are automatically generated in the bottom half of the script, and 
include several individual plots:

+ Taxon_Numbers.pdf
+ Sequence_Lengths.pdf
+ Gaps_Percents.pdf
+ Informative_Sites_Numbers.pdf
+ Informative_Sites_Percents.pdf
+ Missing_Data_Percents.pdf
+ Sequence_Lengths_vs_Number_Informative_Sites.pdf

In addition, a file is created with multiple plots from above included:

+ Taxa_SeqLength_MissData_PercInfSites_LvI.pdf



**Citation Information:**

These scripts were published as part of the following work:

+ *Portik, D.M., Smith, L.L., and K. Bi. 2016. An evaluation of transcriptome-based exon capture for frog phylogenomics across multiple scales of divergence (Class: Amphibia, Order: Anura). Molecular Ecology Resources 16: 1069-1083. https://doi.org/10.1111/1755-0998.12541*

If you use or modify this script for your own purposes, please cite this publication.


**Contact:**

Daniel Portik, PhD

Postdoctoral Researcher

University of Arizona

daniel.portik@gmail.com


